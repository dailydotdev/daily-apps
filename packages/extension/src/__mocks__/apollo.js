import { createMockClient } from 'mock-apollo-client';

// eslint-disable-next-line import/prefer-default-export
export const apolloClient = createMockClient();
