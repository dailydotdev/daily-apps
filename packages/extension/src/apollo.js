import { InMemoryCache } from 'apollo-cache-inmemory';
import { CachePersistor } from 'apollo-cache-persist';
import { createPersistedQueryLink } from 'apollo-link-persisted-queries';
import { createHttpLink } from 'apollo-link-http';
import { ApolloClient } from 'apollo-client';

const cache = new InMemoryCache();

export const persistor = new CachePersistor({
  cache,
  key: 'apollo-cache-persistor',
  storage: {
    async getItem(key) {
      const res = await browser.storage.local.get(key);
      if (!(Number.isNaN(res[key]) || res[key] === undefined)) {
        return res[key];
      }
      return undefined;
    },
    setItem: (key, data) => browser.storage.local.set({ [key]: data }),
    removeItem: key => browser.storage.local.remove(key),
  },
});

const link = createPersistedQueryLink({ useGETForHashedQueries: true })
  .concat(createHttpLink({
    uri: `${process.env.VUE_APP_API_URL}/graphql`,
    useGETForQueries: true,
    credentials: 'include',
  }));

export const apolloClient = new ApolloClient({
  link,
  cache,
});
