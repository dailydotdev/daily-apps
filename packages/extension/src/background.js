import { authService } from './common/services';
import {
  setCache,
  getCache,
  ANALYTICS_ID_KEY,
  DATA_VERSION_KEY,
  CURRENT_DATA_VERSION,
  STATE_KEY,
} from './common/cache';

const cacheUserId = async () => {
  const profile = await authService.getUserProfile();
  return setCache(ANALYTICS_ID_KEY, profile.id);
};

const migrateVersion1 = async () => {
  const [bookmarks, pubs, settings, profile, notificationTime] = await Promise.all([
    getCache('bookmarks', []),
    getCache('publications', {}),
    getCache('settings', {}),
    getCache('profile', null),
    getCache('notificationTime', null),
  ]);
  return {
    feed: {
      bookmarks,
      publications: Object.keys(pubs).map(id => ({ id, enabled: pubs[id] })),
    },
    ui: {
      lastNotificationTime: notificationTime,
      ...settings,
    },
    user: { profile },
  };
};

const migrateCache = async () => {
  const version = await getCache(DATA_VERSION_KEY, '1');
  if (version !== CURRENT_DATA_VERSION) {
    const newCache = await migrateVersion1();
    await setCache(STATE_KEY, newCache);
  }
  await setCache(DATA_VERSION_KEY, CURRENT_DATA_VERSION);
};

browser.browserAction.onClicked.addListener(() => {
  const url = browser.extension.getURL('index.html?source=button');
  browser.tabs.create({ url, active: true });
});

browser.alarms.create('cacheUserId', { periodInMinutes: 60 });

browser.alarms.onAlarm.addListener(async (alarm) => {
  if (alarm.name === 'cacheUserId') {
    // TODO: handle error
    await cacheUserId();
  }
});

browser.runtime.onInstalled.addListener(async () => {
  await Promise.all([
    cacheUserId(),
    migrateCache(),
  ]);
});
