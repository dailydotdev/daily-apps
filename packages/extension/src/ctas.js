import { browserName } from './common/browser';

const ctas = [{
  text: 'Daily is 100% open source',
  icon: 'github',
  link: 'https://github.com/dailydotdev/daily',
  name: 'Github',
  style: {
    color: 'white',
    background: 'black',
  },
}, {
  text: 'Follow us on Instagram',
  logo: 'instagram',
  link: 'https://www.instagram.com/dailydotdev/',
  name: 'Instagram',
  style: {
    color: '#838886',
    background: 'white',
  },
}, {
  text: 'Join our Slack',
  logo: 'slack',
  link: 'https://slack.daily.dev',
  name: 'Slack',
  style: {
    color: '#838886',
    background: 'white',
  },
}, {
  text: 'Follow real-time updates',
  icon: 'twitter',
  link: 'https://twitter.com/dailydotdev',
  name: 'Twitter',
  style: {
    color: 'white',
    background: '#00a2f9',
  },
}, {
  text: 'Like us on Facebook',
  icon: 'facebook',
  link: 'https://www.facebook.com/dailydotdev',
  name: 'Facebook',
  style: {
    color: 'white',
    background: '#3B5998',
  },
}];

if (browserName === 'chrome') {
  ctas.push({
    text: 'We need your review :)',
    logo: 'chrome_store',
    link: 'https://chrome.google.com/webstore/detail/daily-discover-web-techno/jlmpjdjjbgclbocgajdjefcidcncaied/reviews',
    name: 'Chrome',
    style: {
      color: '#838886',
      background: 'white',
    },
  });
} else if (browserName === 'firefox') {
  ctas.push({
    text: 'We need your review :)',
    logo: 'firefox',
    link: 'https://addons.mozilla.org/en-US/firefox/addon/daily/reviews/',
    name: 'Firefox',
    style: {
      color: 'white',
      background: '#0e1028',
    },
  });
}

export default ctas;
