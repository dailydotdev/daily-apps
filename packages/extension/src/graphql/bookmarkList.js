import gql from 'graphql-tag';

export const CREATE_BOOKMARK_LIST_MUTATION = gql`
mutation CreateBookmarkList($name: String!) {
  createBookmarkList(name: $name) {
    id, name
  }
}`;

export const RENAME_BOOKMARK_LIST_MUTATION = gql`
mutation RenameBookmarkList($id: ID!, $name: String!) {
  renameBookmarkList(id: $id, name: $name) {
    id, name
  }
}`;

export const REMOVE_BOOKMARK_LIST_MUTATION = gql`
mutation RemoveBookmarkList($id: ID!) {
  removeBookmarkList(id: $id) {
    _
  }
}`;

export const BOOKMARK_LISTS_QUERY = gql`
query BookmarkLists {
  bookmarkLists {
    id, name
  }
}`;
