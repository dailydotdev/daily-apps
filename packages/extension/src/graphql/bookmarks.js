import gql from 'graphql-tag';

export const ADD_BOOKMARKS_MUTATION = gql`
mutation AddBookmarks($data: AddBookmarkInput!) {
    addBookmarks(data: $data) {
        _
    }
}`;

export const ADD_BOOKMARK_TO_LIST_MUTATION = gql`
mutation AddBookmarkToList($id: ID!, $listId: ID) {
    addBookmarkToList(id: $id, listId: $listId) {
        _
    }
}`;

export const REMOVE_BOOKMARK_MUTATION = gql`
mutation RemoveBookmark($id: ID!) {
    removeBookmark(id: $id) {
        _
    }
}`;
