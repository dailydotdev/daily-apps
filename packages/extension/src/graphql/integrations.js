import gql from 'graphql-tag';

export const POPULAR_INTEGRATIONS_QUERY = gql`
query PopularIntegrations {
  popularIntegrations {
    logo, title, subtitle, url
  }
}`;

export const RSS_FEEDS_QUERY = gql`
query RSSFeeds {
  rssFeeds {
    name, url
  }
}`;
