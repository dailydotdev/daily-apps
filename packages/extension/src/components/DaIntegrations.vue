<template>
  <da-modal class="bookmark-modal integrations" @close="$emit('close')">
    <button class="btn-icon modal__close-btn" @click="$emit('close')">
      <svgicon name="x"/>
    </button>
    <div class="integrations__header">
      <h3>Integrate Daily with +100 apps</h3>
      <p>Choose the feed you want to use for integration</p>
      <da-dropdown v-if="feeds.length" :items="feeds" :selected-index="selectedFeed"
        @selected="selectedFeed = $event" />
      <da-svg src="/graphics/premium.svg" class="premium-badge"/>
    </div>
    <div class="integrations__footer">
      <da-text-field v-if="isPremium" ref="url" class="integration__url-box" label="Feed URL"
                  hide-label :value="selectedFeedUrl" autofocus readonly>
        <template v-slot:right>
          <button class="btn-icon" v-tooltip.top.ignoreClick="copyTooltip"
                  @click="copyFeed" @mouseenter="copyTooltip = 'Copy'">
            <svgicon name="copy"/>
          </button>
        </template>
      </da-text-field>
      <button v-else class="btn btn-big btn-invert integrations__btn" autofocus
              @click="showPremium">
        Show me the link
      </button>
      <p>
        Copy the link above and use it to integrate with +100 awesome apps!
      </p>
      <div class="integration__list">
        <a v-for="item in popularIntegrations" :key="item.title"
            class="integration__list-item" :href="item.url" target="_blank">
          <img :src="item.logo" />
          <div>
            <span class="lil2">{{ item.title }}</span>
            <span class="micro2">{{ item.subtitle }}</span>
          </div>
        </a>
      </div>
    </div>
  </da-modal>
</template>

<script>
import { mapMutations, mapGetters } from 'vuex';
import DaModal from '@daily/components/src/components/DaModal.vue';
import DaDropdown from '@daily/components/src/components/DaDropdown.vue';
import DaTextField from '@daily/components/src/components/DaTextField.vue';
import '@daily/components/icons/x';
import '@daily/components/icons/copy';
import DaSvg from './DaSvg.vue';
import { POPULAR_INTEGRATIONS_QUERY, RSS_FEEDS_QUERY } from '../graphql/integrations';

export default {
  name: 'DaIntegrations',

  components: {
    DaModal,
    DaDropdown,
    DaTextField,
    DaSvg,
  },

  apollo: {
    popularIntegrations: {
      query: POPULAR_INTEGRATIONS_QUERY,
      fetchPolicy: 'cache-and-network',
    },
    rssFeeds: {
      query: RSS_FEEDS_QUERY,
      fetchPolicy: 'cache-and-network',
      skip() {
        return !this.isPremium;
      },
    },
  },

  data() {
    return {
      copyTooltip: 'Copy',
      rssFeeds: [
        { url: 'https://placeholder.com', name: 'News feed' },
        { url: 'https://placeholder2.com', name: 'Bookmarks' },
      ],
      selectedFeed: 0,
    };
  },

  computed: {
    feeds() {
      return this.rssFeeds.map(f => ({ value: f.url, text: f.name }));
    },
    selectedFeedUrl() {
      if (this.feeds.length) {
        return this.feeds[this.selectedFeed].value;
      }
      return '';
    },
    ...mapGetters('user', ['isPremium']),
  },

  watch: {
    async selectedFeedUrl() {
      await this.$nextTick();
      if (this.$refs.url) {
        const copy = this.$refs.url.$el.querySelector('input');
        copy.setSelectionRange(0, 0);
      }
    },
  },

  methods: {
    copyFeed() {
      const copy = this.$refs.url.$el.querySelector('input');
      copy.select();
      copy.setSelectionRange(0, 99999);
      document.execCommand('copy');
      copy.setSelectionRange(0, 0);
      this.copyTooltip = 'Copied';
    },

    showPremium() {
      this.$emit('close');
      setTimeout(() => {
        this.setShowPremium(true);
      }, 50);
    },

    ...mapMutations({
      setShowPremium: 'ui/setShowPremium',
    }),
  },
};
</script>

<style>
.integrations .modal__container {
  padding: 0;
  width: 500px;
  align-items: stretch;

  & .integrations__header p {
    max-width: none;
  }

  & .integration__url-box {
    margin: 0 0 8px;
  }

  & .dropdown {
    width: 360px;
    margin-top: 8px;
  }

  & .premium-badge {
    position: absolute;
    left: 0;
    right: 0;
    bottom: -9px;
    margin: 0 auto;
  }
}

.integrations__header {
  position: relative;
  display: flex;
  flex-direction: column;
  align-items: center;
  padding: 40px 40px 32px;
  background: var(--theme-background-secondary);
}

.integrations__footer {
  display: flex;
  flex-direction: column;
  align-items: center;
  padding: 32px;
  border-top: solid 1px var(--theme-shine);
}

.integration__list {
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  align-self: stretch;
  margin: 8px -8px -8px;
}

.integration__list-item {
  display: flex;
  width: 210px;
  flex-direction: row;
  align-items: center;
  margin: 8px;
  padding: 12px;
  border: 1px solid var(--theme-shine);
  border-radius: 8px;

  & img {
    width: 32px;
    height: 32px;
    border-radius: 8px;
    margin-right: 12px;
  }

  & > div {
    display: flex;
    flex-direction: column;
  }

  & span {
    color: var(--theme-secondary);
    text-transform: none;

    &:first-child {
      color: var(--theme-primary);
    }
  }
}

.integrations__btn {
  width: 210px;
  margin-bottom: 24px;
  justify-content: center;
}
</style>
