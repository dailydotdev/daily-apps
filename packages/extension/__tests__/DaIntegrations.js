import { mount, createLocalVue } from '@vue/test-utils';
import Vuex from 'vuex';
import VueApollo from 'vue-apollo';
import icons from '@daily/components/src/icons';
import tooltip from '@daily/components/src/directives/tooltip';
import DaModal from '@daily/components/src/components/DaModal.vue';
import DaTextField from '@daily/components/src/components/DaTextField.vue';
import DaDropdown from '@daily/components/src/components/DaDropdown.vue';
import { apolloClient } from '../src/apollo';
import DaIntegrations from '../src/components/DaIntegrations.vue';
import { POPULAR_INTEGRATIONS_QUERY, RSS_FEEDS_QUERY } from '../src/graphql/integrations';

jest.mock('../src/apollo');

let localVue = createLocalVue();

localVue.use(Vuex);
localVue.use(VueApollo);
localVue.use(icons);
localVue.directive('tooltip', tooltip(localVue));
localVue.component('da-modal', DaModal);
localVue.component('da-text-field', DaTextField);
localVue.component('da-dropdown', DaDropdown);

const popularIntegrationsHandler = jest.fn();
const rssFeedsHandler = jest.fn();
apolloClient.setRequestHandler(POPULAR_INTEGRATIONS_QUERY, popularIntegrationsHandler);
apolloClient.setRequestHandler(RSS_FEEDS_QUERY, rssFeedsHandler);

let user;
let ui;
let store;

beforeEach(() => {
  window.ga = () => {
  };

  popularIntegrationsHandler.mockReset();
  popularIntegrationsHandler.mockResolvedValue({data: { popularIntegrations: [
    { logo: 'https://a.com/logo', title: 'A', subtitle: 'a', url: 'https://a.com' },
    { logo: 'https://b.com/logo', title: 'B', subtitle: 'b', url: 'https://b.com' },
  ]}});
  rssFeedsHandler.mockReset();
  rssFeedsHandler.mockResolvedValue({data: { rssFeeds: [
    { name: 'News feed', url: 'https://daily.dev/rss/news' },
    { name: 'Bookmarks', url: 'https://daily.dev/rss/bookmarks' },
  ]}});

  user = {
    namespaced: true,
    state: {
      profile: { premium: true },
    },
    getters: {
      isPremium: state => !!state.profile && !!state.profile.premium,
    },
  };

  ui = {
    namespaced: true,
    state: {},
    mutations: {
      setShowPremium: jest.fn(),
    },
  };

  store = new Vuex.Store({
    modules: { user, ui },
  });
});

it('should emit close on close button click', () => {
  const wrapper = mount(DaIntegrations, { store, localVue, apolloProvider: new VueApollo({defaultClient: apolloClient }) });
  wrapper.find('.modal__close-btn').trigger('click');
  expect(wrapper.emitted().close[0]).toBeTruthy();
});

it('should show popular integrations', async () => {
  const wrapper = mount(DaIntegrations, { store, localVue, apolloProvider: new VueApollo({defaultClient: apolloClient }) });
  await wrapper.vm.$nextTick();
  expect(wrapper.find('.integration__list').element).toMatchSnapshot();
});

it('should populate rss feeds in dropdown', async () => {
  const wrapper = mount(DaIntegrations, { store, localVue, apolloProvider: new VueApollo({defaultClient: apolloClient }) });
  await wrapper.vm.$nextTick();
  expect(wrapper.vm.feeds).toMatchSnapshot();
});

it('should set url field value to first rss', async () => {
  const wrapper = mount(DaIntegrations, { store, localVue, apolloProvider: new VueApollo({defaultClient: apolloClient }) });
  await wrapper.vm.$nextTick();
  expect(wrapper.find('.integration__url-box input').element.value).toEqual('https://daily.dev/rss/news');
});

it('should update url field value according to selection', async () => {
  const wrapper = mount(DaIntegrations, { store, localVue, apolloProvider: new VueApollo({defaultClient: apolloClient }) });
  await wrapper.vm.$nextTick();
  wrapper.find('.dropdown').vm.$emit('selected', 1);
  expect(wrapper.find('.integration__url-box input').element.value).toEqual('https://daily.dev/rss/bookmarks');
});

it('should not show premium button when premium user', async () => {
  const wrapper = mount(DaIntegrations, { store, localVue, apolloProvider: new VueApollo({defaultClient: apolloClient }) });
  await wrapper.vm.$nextTick();
  expect(wrapper.find('.integrations__btn').element).toBeFalsy();
});

it('should show premium popup when clicking on the premium button', async (done) => {
  user.state.profile.premium = false;
  const wrapper = mount(DaIntegrations, { store, localVue, apolloProvider: new VueApollo({defaultClient: apolloClient }) });
  await wrapper.vm.$nextTick();
  wrapper.find('.integrations__btn').trigger('click');
  setTimeout(() => {
    expect(wrapper.emitted().close[0]).toBeTruthy();
    expect(ui.mutations.setShowPremium).toBeCalledWith(expect.anything(), true);
    done();
  }, 100);
});