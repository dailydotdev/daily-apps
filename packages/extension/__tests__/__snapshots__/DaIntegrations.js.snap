// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should populate rss feeds in dropdown 1`] = `
Array [
  Object {
    "text": "News feed",
    "value": "https://daily.dev/rss/news",
  },
  Object {
    "text": "Bookmarks",
    "value": "https://daily.dev/rss/bookmarks",
  },
]
`;

exports[`should show popular integrations 1`] = `
<div
  class="integration__list"
>
  <a
    class="integration__list-item"
    href="https://a.com"
    rel="noopener noreferrer"
    target="_blank"
  >
    <img
      src="https://a.com/logo"
    />
     
    <div>
      <span
        class="lil2"
      >
        A
      </span>
       
      <span
        class="micro2"
      >
        a
      </span>
    </div>
  </a>
  <a
    class="integration__list-item"
    href="https://b.com"
    rel="noopener noreferrer"
    target="_blank"
  >
    <img
      src="https://b.com/logo"
    />
     
    <div>
      <span
        class="lil2"
      >
        B
      </span>
       
      <span
        class="micro2"
      >
        b
      </span>
    </div>
  </a>
</div>
`;
